packages/vm/src/imports.rs

```diff
@@ -90,28 +90,63 @@ pub fn do_remove<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResul
     Ok(())
 }
 
-pub fn do_canonicalize_address<A: Api>(
+pub fn do_canonicalize_address<A: Api, S: Storage, Q: Querier>(
     api: A,
     ctx: &mut Ctx,
     source_ptr: u32,
     destination_ptr: u32,
-) -> VmResult<()> {
-    let human: HumanAddr = read_string_region(ctx, source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?.into();
-    let canon = api.canonical_address(&human)?;
-    write_region(ctx, destination_ptr, canon.as_slice())?;
-    Ok(())
+) -> VmResult<u32> {
+    let source_data = read_region(ctx, source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?;
+    if source_data.is_empty() {
+        return Ok(create_error_region::<S, Q>(ctx, "Input is empty")?);
+    }
+
+    let source_string = match String::from_utf8(source_data) {
+        Ok(s) => s,
+        Err(_) => {
+            return Ok(create_error_region::<S, Q>(
+                ctx,
+                "Input is not valid UTF-8",
+            )?)
+        }
+    };
+    let human: HumanAddr = source_string.into();
+    match api.canonical_address(&human) {
+        Ok(canon) => {
+            write_region(ctx, destination_ptr, canon.as_slice())?;
+            Ok(0)
+        }
+        // This check is unrelyable since we cannot tell by the error type if the error should be
+        // reported to the contract or indicates a broken backend.
+        // Err(FfiError::Other { error, .. }) => Ok(create_error_region::<S, Q>(ctx, &error)?),
+        Err(error) => Err(error.into()),
```

Comment:
To use the code you used in your snippet, we will have to add a variant to `FfiError` called `Unknown`, which would represent the problems we really don't know how to handle, and don't want to report to the contract. I guess that was your intention here?

Reply:
Right. If we have `FfiError::Unknown`, how would you call the error case to report to the user like the line I commented out?

Reply:
I was thinking just `Other` is good, but maybe `Error` would be clearer? would sound repetitive though, to write `FfriError::Error(message)`

Reply:
For this usecase I look for "input error"s as those are the ones worth reporting back to the contract. But I don't know if this is too specific. Maybe "API error" is better.

Reply:
I'm not sure either... I think for now calling it `InputError` is ok, but `ApiError` is more general. perhaps it is the better option.